src/Psl/DateTime/DateTime.php

```diff
@@ -0,0 +1,490 @@
+     *
+     * @mutation-free
+     */
+    public function getMinutes(): int
+    {
+        return $this->minutes;
+    }
+
+    /**
+     * Returns the seconds.
+     *
+     * @return int<0, 59>
+     *
+     * @mutation-free
+     */
+    public function getSeconds(): int
+    {
+        return $this->seconds;
+    }
+
+    /**
+     * Returns the nanoseconds.
+     *
+     * @return int<0, 999999999>
+     *
+     * @mutation-free
+     */
+    public function getNanoseconds(): int
+    {
+        return $this->nanoseconds;
+    }
+
+    /**
+     * Gets the timezone associated with the date and time.
+     *
+     * @mutation-free
+     */
+    public function getTimezone(): Timezone
+    {
+        return $this->timezone;
+    }
+
+    /**
+     * Returns a new instance with the specified date.
+     *
+     * @param Month|int<1, 12> $month
+     * @param int<1, 31> $day
+     *
+     * @throws Exception\InvalidArgumentException If specifying the date would result in an invalid date/time.
+     *                                            This can happen if the combination of year, month, and day does not constitute a valid date (e.g., April 31st, February 29th in a non-leap year).
+     *
+     * @mutation-free
+     */
+    public function withDate(int $year, Month|int $month, int $day): static
+    {
+        return static::fromParts(
+            $this->getTimezone(),
+            $year,
+            $month,
+            $day,
+            $this->getHours(),
+            $this->getMinutes(),
+            $this->getSeconds(),
+            $this->getNanoseconds(),
+        );
+    }
+
+    /**
+     * Returns a new instance with the specified time.
+     *
+     * @param int<0, 23> $hours
+     * @param int<0, 59> $minutes
+     * @param int<0, 59> $seconds
+     * @param int<0, 999999999> $nanoseconds
+     *
+     * @throws Exception\InvalidArgumentException If specifying the time would result in an invalid time (e.g., hours greater than 23, minutes or seconds greater than 59).
+     *
+     * @mutation-free
+     */
+    public function withTime(int $hours, int $minutes, int $seconds = 0, int $nanoseconds = 0): static
+    {
+        return static::fromParts(
+            $this->getTimezone(),
+            $this->getYear(),
+            $this->getMonth(),
+            $this->getDay(),
+            $hours,
+            $minutes,
+            $seconds,
+            $nanoseconds,
+        );
+    }
+
+    /**
+     * Adds the specified duration to this date-time object, returning a new instance with the added duration.
+     *
+     * @throws Exception\UnderflowException If adding the duration results in an arithmetic underflow.
+     * @throws Exception\OverflowException If adding the duration results in an arithmetic overflow.
+     *
+     * @mutation-free
+     */
+    public function plus(Duration $duration): static
+    {
+        return static::fromTimestamp($this->timezone, $this->getTimestamp()->plus($duration));
```

Comment:
Every plus / minus could result in the previous / next month or leap year problematic.
We are now only solving it for months and years - but the same could really go for minutes and hours right?
Not sure what the best place is to handle it.

Reply:
I'm also wondering : how does it deal with daylight saving time?
As in : If I add an hour an half hour before the DST hour get shifted, it should be the same hour.

(for example : 2.30am + 1h = 2.30am when at 3.00am the time gets shifted back to 2.00am) Sure it's a weird scenario which I don't really know if it's actually valid.

Reply:
It deals with it better than PHP i would say :D not perfect maybe?

```php
<?php

declare(strict_types=1);

use Psl\DateTime;
use Psl\IO;

require_once 'vendor/autoload.php';

$timestamp = DateTime\Timestamp::fromParts(seconds: 1711846740);
$datetime = DateTime\DateTime::fromTimestamp($timestamp, DateTime\Timezone::EuropeLondon);

// outputs: 31 Mar 2024, 00:59:00 / 0 second(s) offset
IO\write_error_line('%s / %s offset', $datetime->toString(), $datetime->getTimezoneOffset()->toString());

// adding 1-hour results in entering DST, as it starts 1am
$datetime = $datetime->plus(DateTime\Duration::hours(1));

// outputs: 31 Mar 2024, 02:59:00 / 1 hour(s) offset
IO\write_error_line('%s / %s offset', $datetime->toString(), $datetime->getTimezoneOffset()->toString());

// let's leave DST, and remove the hour we added
$datetime = $datetime->plusMonths(7)->minusHours(1);

// output: 31 Oct 2024, 01:59:00 / 0 second(s) offset
IO\write_error_line('%s / %s offset', $datetime->toString(), $datetime->getTimezoneOffset()->toString());
```

```sh
$ php foo.php
31 Mar 2024, 00:59:00 / 0 second(s) offset
31 Mar 2024, 02:59:00 / 1 hour(s) offset
31 Oct 2024, 01:59:00 / 0 second(s) offset
```

Reply:
Not sure what your question is about. Would you expect the last one to be `1:59 at offset 0`?

I'm really not sure on how it should work without having used it myself. But I remember having stumbled against an issue at some moment where I for example did 'today at 20:00' + 1 week, which resulted in `21:00` or `19:00` which broke my scheduling. (the bug was in a flutter app that was written in [dart](https://medium.com/pinch-nl/datetime-and-daylight-saving-time-in-dart-9c9468633b5d)). That is basically the problem I would not like to face myself. The main edge cases of course lie in the moment of the time-switch from e.g. 3.00 -> 2.00, cause there is 2.00:+1 and 2:00+0 I suppose.

Reply:
I'll test how this works in chrono ( https://docs.rs/chrono/latest/chrono/struct.Date.html ) and other libs, and try to implement the same behavior.